File: cpp/samples/core/gmodtraversal.cpp
/**
 * @file gmodtraversal.cpp
 * @brief Demonstrates usage of vista-sdk Gmod depth-first traversal API
 * @details This sample shows how to traverse the Gmod tree using TraversalHandlerResult
 *          to control traversal flow, collect nodes at a given depth, and stop early
 */

#include <dnv/VistaSDK.h>

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk GmodTraversal Sample ===\n\n";

    {
        std::cout << "1. Gmod::traverse: Counting all nodes depth-first\n";
        std::cout << "--------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::size_t total = 0;
        gmod.traverse(
            [&total]([[maybe_unused]] const TraversalPath& parents, [[maybe_unused]] const GmodNode& node)
                -> TraversalHandlerResult {
                ++total;
                return TraversalHandlerResult::Continue;
            });

        std::cout << "Total nodes visited (DFS): " << total << "\n\n";
    }

    {
        std::cout << "2. Gmod::traverse: Collecting nodes at depth 2\n";
        std::cout << "------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::vector<std::string> depth2;
        gmod.traverse([&depth2](const TraversalPath& parents, const GmodNode& node) -> TraversalHandlerResult {
            if (parents.size() == 2)
            {
                depth2.push_back(std::string(node.code()));
            }
            return TraversalHandlerResult::Continue;
        });

        std::cout << "Nodes at depth 2: " << depth2.size() << "\n";
        constexpr std::size_t maxDisplay = 5;
        for (std::size_t i = 0; i < std::min(depth2.size(), maxDisplay); ++i)
        {
            std::cout << "  " << depth2[i] << "\n";
        }
        if (depth2.size() > maxDisplay)
        {
            std::cout << "  ... and " << (depth2.size() - maxDisplay) << " more\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "3. Gmod::traverse: Stopping early after 10 nodes\n";
        std::cout << "-------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::vector<std::string> visited;
        const bool completed = gmod.traverse(
            [&visited]([[maybe_unused]] const TraversalPath& parents, const GmodNode& node) -> TraversalHandlerResult {
                visited.push_back(std::string(node.code()));
                return visited.size() >= 10 ? TraversalHandlerResult::Stop : TraversalHandlerResult::Continue;
            });

        std::cout << "Traversal completed: " << std::boolalpha << completed << "\n";
        std::cout << "First 10 nodes in DFS order:\n";
        for (const auto& code : visited)
        {
            std::cout << "  " << code << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "4. Gmod::traverse: Skipping a subtree\n";
        std::cout << "---------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        // Count nodes that would be visited without skipping root's first child's subtree
        std::size_t skipped = 0;
        std::string skippedCode;
        bool firstChild = true;

        gmod.traverse([&](const TraversalPath& parents, const GmodNode& node) -> TraversalHandlerResult {
            if (parents.size() == 1 && firstChild)
            {
                firstChild = false;
                skippedCode = std::string(node.code());
                return TraversalHandlerResult::SkipSubtree;
            }
            ++skipped;
            return TraversalHandlerResult::Continue;
        });

        std::cout << "Skipped subtree rooted at: " << skippedCode << "\n";
        std::cout << "Nodes visited after skip : " << skipped << "\n\n";
    }

    {
        std::cout << "5. Gmod::traverse: Using state overload\n";
        std::cout << "----------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        struct State
        {
            std::size_t leafCount = 0;
            std::size_t maxDepth = 0;
        };

        State state;
        Gmod::TraverseHandlerWithState<State> handler =
            [](State& s, const TraversalPath& parents, const GmodNode& node) -> TraversalHandlerResult {
            if (node.isLeafNode())
            {
                ++s.leafCount;
            }
            s.maxDepth = std::max(s.maxDepth, parents.size());
            return TraversalHandlerResult::Continue;
        };
        gmod.traverse(state, handler);

        std::cout << "Leaf nodes visited: " << state.leafCount << "\n";
        std::cout << "Max path depth    : " << state.maxDepth << "\n\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/core/gmod_traversal.rs
use vista_sdk::core::gmod::TraversalHandlerResult;
use vista_sdk::core::vis::Vis;

fn main() {
    println!("=== vista-sdk GmodTraversal Sample ===\n");

    {
        println!("1. Gmod::traverse: Counting all nodes depth-first");
        println!("--------------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let mut total = 0usize;
        gmod.traverse(
            |_parents, _node| {
                total += 1;
                TraversalHandlerResult::Continue
            },
            1,
        );

        println!("Total nodes visited (DFS): {}\n", total);
    }

    {
        println!("2. Gmod::traverse: Collecting nodes at depth 2");
        println!("------------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let mut depth2 = Vec::new();
        gmod.traverse(
            |parents, node| {
                if parents.len() == 2 {
                    depth2.push(node.code().to_string());
                }
                TraversalHandlerResult::Continue
            },
            1,
        );

        println!("Nodes at depth 2: {}", depth2.len());
        for code in depth2.iter().take(5) {
            println!("  {}", code);
        }
        if depth2.len() > 5 {
            println!("  ... and {} more", depth2.len() - 5);
        }

        println!();
    }

    {
        println!("3. Gmod::traverse: Stopping early after 10 nodes");
        println!("-------------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let mut visited = Vec::new();
        let completed = gmod.traverse(
            |_parents, node| {
                visited.push(node.code().to_string());
                if visited.len() >= 10 {
                    TraversalHandlerResult::Stop
                } else {
                    TraversalHandlerResult::Continue
                }
            },
            1,
        );

        println!("Traversal completed: {}", completed);
        println!("First 10 nodes in DFS order:");
        for code in &visited {
            println!("  {}", code);
        }

        println!();
    }

    {
        println!("4. Gmod::traverse: Skipping a subtree");
        println!("---------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let mut skipped_code = String::new();
        let mut first_child = true;
        let mut count = 0usize;

        gmod.traverse(
            |parents, node| {
                if parents.len() == 1 && first_child {
                    first_child = false;
                    skipped_code = node.code().to_string();
                    return TraversalHandlerResult::SkipSubtree;
                }
                count += 1;
                TraversalHandlerResult::Continue
            },
            1,
        );

        println!("Skipped subtree rooted at: {}", skipped_code);
        println!("Nodes visited after skip : {}\n", count);
    }

    {
        println!("5. Gmod::traverse: Accumulating state");
        println!("---------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let mut leaf_count = 0usize;
        let mut max_depth = 0usize;

        gmod.traverse(
            |parents, node| {
                if node.is_leaf_node() {
                    leaf_count += 1;
                }
                max_depth = max_depth.max(parents.len());
                TraversalHandlerResult::Continue
            },
            1,
        );

        println!("Leaf nodes visited: {}", leaf_count);
        println!("Max path depth    : {}\n", max_depth);
    }
}
